### PR TITLE
Handle overflow of validator groups

### LINKED ISSUE #103: Handle overflow of validator groups

### PR DESCRIPTION
**Pull request by pahor167**
### Description

When groups that we are voting for in stCELO protocol are close to it's voting capacity we will overflow votes to default strategy.

### Other changes

No other changes

### Tested

Unit tests + e2e test

### Related issues

- Fixes #103

### CODE REVIEW

File: contracts/Manager.sol [hunk 1]
```diff
@@ -132,6 +132,18 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
...
     error RebalanceEnoughCelo(address group, uint256 actualCelo, uint256 expectedCelo);
 
+    /**
+     * @notice Used when trying to overflow rebalance group that is not overflowing.
+     * @param group The group's address.
+     */
+    error FromGroupNotOverflowing(address group);
+
+    /**
+     * @notice Used when trying to overflow rebalance to group that is overflowing.
```
Comment (soloseng) [inline contracts/Manager.sol:142]:
```suggestion
     * @notice Used when trying to rebalance to a group that is overflowing.
```

File: contracts/Manager.sol [hunk 2]
```diff
@@ -547,6 +603,32 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
         return strategy;
     }
 
+    /**
+     * Returns votes count that can be received by group through stCELO protocol.
+     * @param group The group that can receive votes
+     */
```
Comment (soloseng) [inline contracts/Manager.sol:609]: Add return to NATSPEC

File: contracts/Manager.sol [hunk 3]
```diff
@@ -411,11 +423,52 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
...
+        }
+
+        uint256 fromReceivableVotes = getReceivableVotesForGroup(fromGroup);
+        if (fromReceivableVotes != 0) {
+            revert FromGroupNotOverflowing(fromGroup);
+        }
+
+        uint256 fromScheduledVotes = account.scheduledVotesForGroup(fromGroup);
+        if (fromScheduledVotes == 0) {
+            revert FromGroupNotOverflowing(fromGroup);
```
Comment (soloseng) [inline contracts/Manager.sol:460]: Should this error be `NoScheduledVotes(group)`

File: contracts/Manager.sol [hunk 4]
```diff
@@ -132,6 +132,18 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
...
+     * @notice Used when trying to overflow rebalance group that is not overflowing.
+     * @param group The group's address.
+     */
+    error FromGroupNotOverflowing(address group);
+
+    /**
+     * @notice Used when trying to overflow rebalance to group that is overflowing.
+     * @param group The group's address.
+     */
+    error ToGroupOverflowing(address group);
```
Comment (soloseng) [inline contracts/Manager.sol:145]: nit: 
```suggestion
    error GroupOverflowing(address group);
```

File: contracts/Manager.sol [hunk 5]
```diff
@@ -132,6 +132,18 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
      */
     error RebalanceEnoughCelo(address group, uint256 actualCelo, uint256 expectedCelo);
 
+    /**
+     * @notice Used when trying to overflow rebalance group that is not overflowing.
+     * @param group The group's address.
+     */
+    error FromGroupNotOverflowing(address group);
```
Comment (soloseng) [inline contracts/Manager.sol:139]: nit:
```suggestion
    error GroupNotOverflowing(address group);
```

File: contracts/SpecificGroupStrategy.sol [hunk 6]
```diff
@@ -27,12 +28,24 @@ contract SpecificGroupStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeab
      * @notice stCELO that was cast for specific group strategies,
      * strategy => stCELO amount
      */
-    mapping(address => uint256) private specificGroupStrategyTotalStCeloVotes;
+    mapping(address => uint256) private stCeloInStrategy;
 
     /**
-     * @notice Total stCELO that was voted with on specific group strategies
+     * @notice Total stCELO that was voted with on specific group strategies (including overflows).
+     * To get actuall stCelo in specific strategy it is necessary to subtract `totalStCeloOverflow`.
```
Comment (soloseng) [inline contracts/SpecificGroupStrategy.sol:35]:
```suggestion
     * To get the actual stCelo in specific strategy it is necessary to subtract `totalStCeloOverflow`.
```

File: contracts/SpecificGroupStrategy.sol [hunk 7]
```diff
@@ -27,12 +28,24 @@ contract SpecificGroupStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeab
...
+    uint256 public totalStCeloLocked;
+
+    /**
+     * @notice stCELO that was cast for specific group strategies and overflown to defautl strategy,
+     * strategy => stCELO amount
+     */
+    mapping(address => uint256) private stCeloInStrategyOverflown;
+
+    /**
+     * @notice Total stCelo that was overflown to default strategy
```
Comment (soloseng) [inline contracts/SpecificGroupStrategy.sol:46]:
```suggestion
     * @notice Total stCelo that was overflown to default strategy.
```

File: contracts/SpecificGroupStrategy.sol [hunk 8]
```diff
@@ -27,12 +28,24 @@ contract SpecificGroupStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeab
...
-     * @notice Total stCELO that was voted with on specific group strategies
+     * @notice Total stCELO that was voted with on specific group strategies (including overflows).
+     * To get actuall stCelo in specific strategy it is necessary to subtract `totalStCeloOverflow`.
      */
-    uint256 private totalStCeloInSpecificGroupStrategies;
+    uint256 public totalStCeloLocked;
+
+    /**
+     * @notice stCELO that was cast for specific group strategies and overflown to defautl strategy,
+     * strategy => stCELO amount
```
Comment (soloseng) [inline contracts/SpecificGroupStrategy.sol:41]: period

File: test-ts/manager.test.ts [hunk 9]
```diff
@@ -1031,26 +1065,126 @@ describe("Manager", () => {
...
+
+    describe("when groups are close to their voting limit", () => {
+      const firstGroupCapacity = parseUnits("40.166666666666666666");
+      const depositAmount = parseUnits("50");
+      let originalDefaultHead: string;
+      let previousOfHead: string;
+      let originalOverflow: BigNumber;
+
+      beforeEach(async () => {
+        // For more info about these numbers check comment above
```
Comment (soloseng) [inline test-ts/manager.test.ts:1097]: is this comment still relevant?

File: test-ts/utils.ts [hunk 10]
```diff
@@ -166,7 +168,8 @@ export async function activateValidators(
   multisigOwner: string,
   groupAddresses: string[]
 ) {
-  let nextGroup = ADDRESS_ZERO;
+  let [nextGroup] = await defaultStrategyContract.getActiveGroupsTail();
+  console.log("tail", nextGroup);
```
Comment (soloseng) [inline test-ts/utils.ts:172]: is this still needed?

File: test-ts/utils.ts [hunk 11]
```diff
@@ -445,6 +448,17 @@ export async function activateAndVoteTest(deployerAccountName = "deployer") {
...
 }
 
+export async function revokeTest(deployerAccountName = "deployer") {
+  try {
+    await hre.run(ACCOUNT_REVOKE, {
+      account: deployerAccountName,
+      useNodeAccount: true,
+    });
+  } catch (error) {
+    throw Error(`Activate and vote failed! ${JSON.stringify(error)}`);
```
Comment (soloseng) [inline test-ts/utils.ts:458]: update error message

File: test-ts/utils.ts [hunk 12]
```diff
@@ -758,3 +775,34 @@ export async function prepareOverflow(
...
+  });
+  await sendFundsTx.wait();
+  await impersonateAccount(accountAddress);
+
+  const accountsContract = await hre.kit.contracts.getAccounts();
+  const createAccountTxObject = accountsContract.createAccount();
+  await createAccountTxObject.send({
+    from: accountAddress,
+  });
+  // TODO: once contractkit updated - use just election contract from contractkit
```
Comment (soloseng) [inline test-ts/utils.ts:797]: Is there a PR out for this change? Could it be referenced here?

**Review by soloseng — APPROVED** [review]
🚀

File: test-ts/manager.test.ts [hunk 13]
```diff
@@ -2113,4 +2286,122 @@ describe("Manager", () => {
...
+        );
+      });
+
+      describe("When from group overflowing", () => {
+        const firstGroupCapacity = parseUnits("40.166666666666666666");
+        beforeEach(async () => {
+          await account.setCeloForGroup(groupAddresses[0], firstGroupCapacity.mul(2));
+          // const [tail] = await defaultStrategyContract.getActiveGroupsTail()
+          // expect(tail).to.eq(groupAddresses[0])
+          // await manager.connect(depositor).deposit({value: firstGroupCapacity})
```
Comment (soloseng) [inline test-ts/manager.test.ts:2348]: comment cleanup

File: test-ts/manager.test.ts (hunk 9 above)
Comment (pahor167) [reply test-ts/manager.test.ts:1097]: no, thanks :D

File: test-ts/utils.ts (hunk 12 above)
Comment (pahor167) [reply test-ts/utils.ts:797]: I would prefer to not to put it into this PR

File: test-ts/end-to-end-overflow.test.ts [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+    );
+  });
+
+  const firstGroupCapacity = parseUnits("40.166666666666666666");
+  const secondGroupCapacity = parseUnits("99.25");
+  const thirdGroupCapacity = parseUnits("200.166666666666666666");
+
+  it("deposit, transfer and activate", async () => {
+    for (let i = 0; i < 3; i++) {
+      console.log(`group[${i}] ${groups[i].address}`);
```
Comment (soloseng) [inline test-ts/end-to-end-overflow.test.ts:137]: are these logs needed?

File: contracts/DefaultStrategy.sol [hunk 15]
```diff
@@ -548,11 +581,11 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
     /**
      * @notice Distributes votes by computing the number of votes to be substracted
```
Comment (m-chrzan) [inline contracts/DefaultStrategy.sol:582]: Typo: subtracted

File: contracts/DefaultStrategy.sol [hunk 16]
```diff
@@ -568,17 +601,17 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
         address votedGroup = activeGroups.getHead();
         uint256 groupsIndex;
 
-        for (groupsIndex = 0; groupsIndex < maxGroupCount; groupsIndex++) {
-            if (votesAmount == 0 || votedGroup == address(0)) {
+        while (groupsIndex < maxGroupCount) {
```
Comment (m-chrzan) [inline contracts/DefaultStrategy.sol:604]: Why the change from `for` to `while`?

File: contracts/DefaultStrategy.sol [hunk 17]
```diff
@@ -630,16 +666,22 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
         uint256 groupsIndex;
 
         for (groupsIndex = 0; groupsIndex < maxGroupCount; groupsIndex++) {
-            if (votesAmount == 0 || votedGroup == address(0)) {
+            if (votedGroup == depositGroupToIgnore) {
+                // this scenario happens when overflow is being utilized and
+                // we want to ignore the groups that is overflowing
```
Comment (m-chrzan) [inline contracts/DefaultStrategy.sol:671]: Typo: group

File: contracts/DefaultStrategy.sol [hunk 18]
```diff
@@ -630,16 +666,22 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
         uint256 groupsIndex;
 
         for (groupsIndex = 0; groupsIndex < maxGroupCount; groupsIndex++) {
-            if (votesAmount == 0 || votedGroup == address(0)) {
+            if (votedGroup == depositGroupToIgnore) {
```
Comment (m-chrzan) [inline contracts/DefaultStrategy.sol:669]: Is this necessary? Shouldn't we already have a mechanism anyways that skips overflowed groups?

File: contracts/Manager.sol [hunk 19]
```diff
@@ -406,11 +423,56 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
...
         );
+
+        scheduleRebalanceTransfer(fromGroup, toGroup, toMove);
+    }
+
+    /**
+     * @notice Rebalance according to CELO overflow rather than stCELO ratio.
+     * If one of the groups is overflowing and there are still some votes that
+     * are scheduled for the group, this function allows to transfer these
+     * votes to any active group in protocol that is not overflowing yet.
```
Comment (m-chrzan) [inline contracts/Manager.sol:444]: Should it go to any group, or specifically to the lowest voted group(s), as per the default strategy?

File: contracts/Manager.sol [hunk 20]
```diff
@@ -479,7 +541,12 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
         uint256 stCELOFromDefaultStrategy;
 
         if (isSpecificGroupStrategy) {
-            stCELOFromSpecificStrategy = specificGroupStrategy.stCeloInGroup(group);
+            uint256 overflow;
+            (stCELOFromSpecificStrategy, overflow) = specificGroupStrategy.getStCeloInStrategy(
```
Comment (m-chrzan) [inline contracts/Manager.sol:545]: Naming nit: `stCelo...` (here and elsewhere if applicable).

File: contracts/SpecificGroupStrategy.sol [hunk 21]
```diff
@@ -33,12 +33,25 @@ contract SpecificGroupStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeab
...
-     * @notice Total stCELO that was voted with on specific group strategies.
+     * @notice Total stCELO that was voted with on specific group strategies (including overflows).
+     * @dev To get the actual stCelo in specific strategy
+     * it is necessary to subtract `totalStCeloOverflow`.
      */
-    uint256 public totalStCeloInStrategy;
+    uint256 public totalStCeloLocked;
+
+    /**
+     * @notice stCELO that was cast for specific group strategies and overflown to defautl strategy,
```
Comment (m-chrzan) [inline contracts/SpecificGroupStrategy.sol:46]: Grammar nit: I *think* this should be "overflowed", not "overflown", here and probably most other cases throughout.

File: contracts/SpecificGroupStrategy.sol [hunk 22]
```diff
@@ -337,53 +373,133 @@ contract SpecificGroupStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeab
...
-        votes[0] = celoWithdrawalAmount;
+        if (stCeloWithdrawalAmount > stCeloInStrategy[strategy]) {
+            revert CantWithdrawAccordingToStrategy(strategy);
+        }
 
-        if (stCeloWithdrawalAmount > stCeloInGroup[group]) {
-            revert CantWithdrawAccordingToStrategy(group);
+        uint256 overflowingStCelo = stCeloInStrategyOverflown[strategy];
+        if (overflowingStCelo > 0) {
+            uint256 overflowingCelo = IManager(manager).toCelo(overflowingStCelo);
```
Comment (m-chrzan) [inline contracts/SpecificGroupStrategy.sol:396]: One thing to think about here: if we knew the target strategy here, we could short-circuit if the target strategy was the default, since we don't actually have to withdraw anything from it, just need to update the amount overflowed for this specific strategy.

File: contracts/Account.sol [hunk 23]
```diff
@@ -463,7 +463,7 @@ contract Account is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Managed, I
 
         // Lock up the unlockedCeloForGroup in LockedGold, which increments the
         // non-voting LockedGold balance for this contract.
-        if (toLock > 0) {
+        if (toLock != 0) {
```
[the remark this answers is not in the document]
Comment (pahor167) [reply contracts/Account.sol:466]: I was reading about smart contract gas optimisations and this was one of those. I reverted it for optimization

File: contracts/DefaultStrategy.sol (hunk 16 above)
Comment (pahor167) [reply contracts/DefaultStrategy.sol:604]: It felt to me as more readable - I could switch it back if you dont think that's the case

File: contracts/SpecificGroupStrategy.sol [hunk 24]
```diff
@@ -127,12 +140,6 @@ contract SpecificGroupStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeab
      */
     error NoActiveGroups();
 
-    /**
-     * @notice Used when attempting to deposit when chosen strategy cannot receive any new votes.
-     * @param group The group's address.
-     */
-    error StrategyCannotReceiveVote(address group);
```
[the remark this answers is not in the document]
Comment (pahor167) [reply contracts/SpecificGroupStrategy.sol:134]: I was thinking about it in a way that we can always activate new validator group and rebalance. Since we are lifting group count limit there should always be new healthy validator group.

File: contracts/Manager.sol [hunk 25]
```diff
@@ -406,11 +423,56 @@ contract Manager is UUPSOwnableUpgradeable, UsingRegistryUpgradeable {
...
+        }
+
+        uint256 toReceivableVotes = getReceivableVotesForGroup(toGroup);
+        if (toReceivableVotes == 0) {
+            revert ToGroupOverflowing(toGroup);
+        }
+
+        uint256 fromActualReceivable = getActualReceivableVotes(fromGroup);
+
+        if (fromActualReceivable >= fromScheduledVotes) {
```
[the remark this answers is not in the document]
Comment (pahor167) [reply contracts/Manager.sol:470]: I refactored the code

File: contracts/DefaultStrategy.sol (hunk 16 above)
Comment (pahor167) [reply contracts/DefaultStrategy.sol:604]: I did refactor it a bit more

File: contracts/DefaultStrategy.sol (hunk 18 above)
Comment (pahor167) [reply contracts/DefaultStrategy.sol:669]: The problem is when group fulfils following conditions:

- strategy is specific strategy
- strategy is active strategy and also is the tail of default strategy
- we are depositing to this group through specific strategy and group will overflow with this deposit

In this scenario we will schedule as much votes as possible to this group through specific strategy and then we will request default strategy to schedule rest. Since this group is also tail of default strategy it will try to schedule votes to this group since votes from specific strategy are not yet really scheduled.

File: contracts/SpecificGroupStrategy.sol (hunk 22 above)
Comment (pahor167) [reply contracts/SpecificGroupStrategy.sol:396]: This will actually happen on Account level. When there will be some votes to be scheduled and some votes to be revoked for same validator group, it will just cancel itself out.

Since I didn't consider this scenario too common it felt like having this mechanism on Account level is sufficient. User will have to pay a bit more for gas fees but protocol will not be hurt in anyway.
